tests/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Source.Tests/DiagnosticAnalysisRunnerTests.cs

```diff
@@ -0,0 +1,116 @@
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.Extensions.Logging;
+using Moq;
+using Xunit;
+
+namespace Microsoft.DotNet.UpgradeAssistant.Steps.Source.Tests
+{
+    public class DiagnosticAnalysisRunnerTests
+    {
+        private readonly Fixture _fixture;
+
+        public DiagnosticAnalysisRunnerTests()
+        {
+            _fixture = new Fixture();
+        }
+
+        [Fact]
+        public async Task DiagnosticsWithEmptyProjectTest()
+        {
+            // Arrange
+            using var mock = AutoMock.GetLoose();
+            var loggerMock = new Mock<ILogger<RoslynDiagnosticProvider>>();
+
+            var project = CreateProject(mock);
+            project.Setup(p => p.GetRoslynProject()).Returns(() =>
+            {
+                var ws = new AdhocWorkspace();
+                var name = Path.GetFileNameWithoutExtension("TestProject.csproj")!;
+                return ws.AddProject(ProjectInfo.Create(ProjectId.CreateNewId(), VersionStamp.Default, name, name, "C#", filePath: "TestProject.csproj"));
+            });
+
+            var context = mock.Mock<IUpgradeContext>();
+            context.Setup(c => c.Projects).Returns(new[] { project.Object });
+            var diagnosticAnalysisRunner = new RoslynDiagnosticProvider(GetAnalyzers(false), Array.Empty<AdditionalText>(), loggerMock.Object);
+
+            // Act
+            var diagnostics = await diagnosticAnalysisRunner.GetDiagnosticsAsync(project.Object, CancellationToken.None).ConfigureAwait(false);
+
+            // Assert
+            Assert.Empty(diagnostics);
+        }
+
+        [Theory]
+        [InlineData(false, 0, null)]
+        [InlineData(true, 1, "Test1")]
+        public async Task DiagnosticsWithProjectFilesTest(bool includeLocation, int expectedDiagnosticCount, string? expectedDiagnosticId)
+        {
+            // Arrange
+            using var mock = AutoMock.GetLoose();
+            var loggerMock = new Mock<ILogger<RoslynDiagnosticProvider>>();
+
+            var project = CreateProject(mock);
+            project.Setup(p => p.GetRoslynProject()).Returns(() =>
+            {
+                var ws = new AdhocWorkspace();
+                var name = Path.GetFileNameWithoutExtension("TestProject.csproj")!;
+                var projectInfo = ProjectInfo.Create(ProjectId.CreateNewId(), VersionStamp.Default, name, name, "C#", filePath: "TestProject.csproj");
+                var documents = new List<DocumentInfo>() { DocumentInfo.Create(DocumentId.CreateNewId(projectInfo.Id), "Foo.cs") };
+                var proj = ws.AddProject(projectInfo.WithDocuments(documents));
+                return proj;
+            });
+
+            var context = mock.Mock<IUpgradeContext>();
+            context.Setup(c => c.Projects).Returns(new[] { project.Object });
+            var diagnosticAnalysisRunner = new RoslynDiagnosticProvider(GetAnalyzers(includeLocation), Array.Empty<AdditionalText>(), loggerMock.Object);
```

Comment:
You should call `mock.Create<RoslynDiagnosticProvider>()` here instead of newing it up yourself.

Reply:
As I mentioned before, I did try mocking the `RoslynDiagnosticProvider` but it was unable to take in parameters like Analyzers, AdditionalTexts and Logger - `System.NullReferenceException : Object reference not set to an instance of an object.
` 
So I decided to new up instead.